### PR TITLE
CI: enable libfuzzer (fixing build and docker)

### PR DESCRIPTION
**Pull request by yakov-olkhovskiy**
### Changelog category (leave one):
- CI Fix or Improvement (changelog entry is not required)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [x]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [x]  Exclude AST fuzzers

### CODE REVIEW

File: src/AggregateFunctions/fuzzers/aggregate_function_state_deserialization_fuzzer.cpp [hunk 1]
```diff
@@ -12,10 +12,33 @@
...
+    extern const int NOT_IMPLEMENTED;
+}
+
+class IFunctionBase;
+using FunctionBasePtr = std::shared_ptr<const IFunctionBase>;
+
+FunctionBasePtr createFunctionBaseCast(
+    ContextPtr, const char *, const ColumnsWithTypeAndName &, const DataTypePtr &, std::optional<CastDiagnostic>, CastType)
+{
+    throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Type conversions are not implemented for Library Bridge");
```
Comment (alexey-milovidov) [inline src/AggregateFunctions/fuzzers/aggregate_function_state_deserialization_fuzzer.cpp:36]: But it isn't a library bridge :)

File: src/Parsers/fuzzers/codegen_fuzzer/codegen_select_fuzzer.cpp [hunk 2]
```diff
@@ -27,7 +27,7 @@ DEFINE_BINARY_PROTO_FUZZER(const Sentence& main)
         DB::ParserQueryWithOutput parser(input.data() + input.size());
         try
         {
-            DB::ASTPtr ast = parseQuery(parser, input.data(), input.data() + input.size(), "", 0, 0);
+            DB::ASTPtr ast = parseQuery(parser, input.data(), input.data() + input.size(), "", 0, 0, 0);
```
Comment (alexey-milovidov) [inline src/Parsers/fuzzers/codegen_fuzzer/codegen_select_fuzzer.cpp:30]: A non-zero value is actually needed to prevent slowness of the fuzzer.

File: src/Parsers/fuzzers/create_parser_fuzzer.cpp [hunk 3]
```diff
@@ -14,7 +14,7 @@ extern "C" int LLVMFuzzerTestOneInput(const uint8_t * data, size_t size)
         std::string input = std::string(reinterpret_cast<const char*>(data), size);
 
         DB::ParserCreateQuery parser;
-        DB::ASTPtr ast = parseQuery(parser, input.data(), input.data() + input.size(), "", 0, 1000);
+        DB::ASTPtr ast = parseQuery(parser, input.data(), input.data() + input.size(), "", 0, 1000, 0);
```
Comment (alexey-milovidov) [inline src/Parsers/fuzzers/create_parser_fuzzer.cpp:17]: A non-zero value is actually needed to prevent slowness of the fuzzer.

File: src/AggregateFunctions/fuzzers/aggregate_function_state_deserialization_fuzzer.cpp (hunk 1 above)
Comment (yakov-olkhovskiy) [reply src/AggregateFunctions/fuzzers/aggregate_function_state_deserialization_fuzzer.cpp:36]: will fix, it's just a fast forward to return it to working state

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit 3c2915934f31d82176b65e1e998eca030671d872 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/61908/3c2915934f31d82176b65e1e998eca030671d872/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Comment by PedroTadim** [comment]
Have you tried to run `codegen_select_fuzzer`? There is a `finalize` call missing for `DB::WriteBufferFromOStream`.

File: tests/fuzz/runner.py [hunk 4]
```diff
@@ -11,7 +11,7 @@
 
 
 def run_fuzzer(fuzzer: str):
-    logging.info(f"Running fuzzer {fuzzer}...")
+    logging.info("Running fuzzer %s...", fuzzer)
```
Comment (rschu1ze) [inline tests/fuzz/runner.py:14]: General comment: to the average developer, it is unclear (at a high level)
- which fuzzers currently exist in ClickHouse,
- how they work each (e.g. do they provide generate random SQL, file data, network input, etc. - based on which library exactly?)
- what steps are needed to run them locally (e.g. which options are needed for CMake, which scripts or tests to call per fuzzer, how do they interact with sanitizers (e.g. below I find *SAN_OPTIONS)
- how are `tests/fuzz/*_fuzzer.options` files related and if/how/when developers should change them
- what fuzzing tests currently run in CI and where to look / how to reproduce when a failure was found.

In other words, we would need a README similar to tests/integration/README.md.

EDIT: I just found https://clickhouse.com/docs/en/development/tests#fuzzing and tests/fuzz/README.md. Anyways, both are incomplete and quite messy. It would be nice to consolidate the information into one comprehensive README.md file (in the repository) and then link from the website to that README (one sentence in the "Fuzzing" section is enough).

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 5]
```diff
@@ -59,6 +59,9 @@ At a minimum, the following information should be added (but add more as needed)
 - [ ] <!---ci_exclude_tsan|msan|ubsan|coverage--> Exclude: All with TSAN, MSAN, UBSAN, Coverage
 - [ ] <!---ci_exclude_aarch64|release|debug--> Exclude: All with aarch64, release, debug
 ---
+- [ ] <!---ci_include_fuzzer--> Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
+- [ ] <!---ci_exclude_ast--> Exclude AST fuzzers
```
Comment (rschu1ze) [inline .github/PULL_REQUEST_TEMPLATE.md:63]: Can we keep the style of existing descriptions above? `Exclude: AST fuzzers`

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 6]
```diff
@@ -59,6 +59,9 @@ At a minimum, the following information should be added (but add more as needed)
 - [ ] <!---ci_exclude_tsan|msan|ubsan|coverage--> Exclude: All with TSAN, MSAN, UBSAN, Coverage
 - [ ] <!---ci_exclude_aarch64|release|debug--> Exclude: All with aarch64, release, debug
 ---
+- [ ] <!---ci_include_fuzzer--> Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
```
Comment (rschu1ze) [inline .github/PULL_REQUEST_TEMPLATE.md:62]: `ci_include_fuzzer` sounds like "run the standard tests and _additionally_ run fuzzers" but the description says "run _only_ fuzzers". This is slightly confusing.

File: tests/fuzz/runner.py (hunk 4 above)
Comment (rschu1ze) [reply tests/fuzz/runner.py:14]: While at it (but I am not sure if you are the right person here) - similar knowledge is ideally documented for OSS fuzz builds which @nickitat [recently fixed](https://github.com/ClickHouse/ClickHouse/pull/67235)). To someone not familiar with fuzzing, it is not clear
- what OSS fuzz tests,
- where fuzzing results can be found,
- how that is different from what we test in our CI,
- who is when and how notified if something goes wrong.

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 6 above)
Comment (yakov-olkhovskiy) [reply .github/PULL_REQUEST_TEMPLATE.md:62]: option's names predicates `ci_include_/ci_exclude_` are dictated how CI works - they automatically disable/enable corresponding jobs

File: tests/fuzz/runner.py (hunk 4 above)
Comment (yakov-olkhovskiy) [reply tests/fuzz/runner.py:14]: I agree that we need more comprehensive guide for libFuzzers, but let's do it separately - our goal here to fix build and make it runnable in our CI - it will help to keep it alive.

**Review by rschu1ze — APPROVED** [review]
